db/migrate/20160527140046_foreign_keys.rb

```diff
@@ -0,0 +1,116 @@
+class ForeignKeys < ActiveRecord::Migration
+  class ClassificationSubject < ActiveRecord::Base
+  end
+
+  def up
+    add_foreign_key :access_control_lists, :user_groups, on_update: :cascade, on_delete: :cascade
+
+    Aggregation.joins("LEFT OUTER JOIN workflows ON workflows.id = aggregations.workflow_id").where("workflows.id IS NULL").delete_all
+    add_foreign_key :aggregations, :workflows, on_update: :cascade, on_delete: :cascade
+    add_foreign_key :aggregations, :subjects, on_update: :cascade, on_delete: :cascade
+
+    add_foreign_key :authorizations, :users, on_update: :cascade, on_delete: :cascade
+
+    # Classification foreign keys are done all the way at the end because we might be deleting projects or workflows
+    # that they refer to on the following lines. Apart from that, everything is in alphabetical order, same as the structure.sql
+
+    # Commented out add_foreign_key calls are ones that already exist. Left them in here so that it's
+    # easier to check if I caught all of them.
+    #
+    # add_foreign_key :classification_subjects, :classifications
+    # add_foreign_key :classification_subjects, :subjects
+
+    add_foreign_key :collections_subjects, :subjects, on_update: :cascade, on_delete: :restrict
+    add_foreign_key :collections_subjects, :collections, on_update: :cascade, on_delete: :restrict
+
+    add_foreign_key :field_guides, :projects, on_update: :cascade, on_delete: :cascade
+
+    add_foreign_key :memberships, :user_groups, on_update: :cascade, on_delete: :cascade
+    add_foreign_key :memberships, :users, on_update: :cascade, on_delete: :cascade
+
+    add_foreign_key :oauth_access_grants, :users, column: :resource_owner_id, on_update: :cascade, on_delete: :cascade
+    add_foreign_key :oauth_access_grants, :oauth_applications, column: :application_id, on_update: :cascade, on_delete: :cascade
+
+    add_foreign_key :oauth_access_tokens, :users, column: :resource_owner_id, on_update: :cascade, on_delete: :cascade
+    add_foreign_key :oauth_access_tokens, :oauth_applications, column: :application_id, on_update: :cascade, on_delete: :cascade
+
+    ProjectContent.joins("LEFT OUTER JOIN projects ON projects.id = project_contents.project_id").where("projects.id IS NULL").delete_all
+    add_foreign_key :project_contents, :projects, on_update: :cascade, on_delete: :cascade
+
+    ProjectPage.joins("LEFT OUTER JOIN projects ON projects.id = project_pages.project_id").where("projects.id IS NULL").delete_all
+    add_foreign_key :project_pages, :projects, on_update: :cascade, on_delete: :cascade
+
+    # add_foreign_key :recents, :classifications
+    # add_foreign_key :recents, :subjects
+
+    SetMemberSubject.joins("LEFT OUTER JOIN subject_sets ON subject_sets.id = set_member_subjects.subject_set_id").where("subject_sets.id IS NULL").delete_all
+    SetMemberSubject.joins("LEFT OUTER JOIN subjects ON subjects.id = set_member_subjects.subject_id").where("subjects.id IS NULL").delete_all
```

Comment:
Out of interest were there many of these kicking around? These are deleted subjects that have orphaned SMS's right?